### PR TITLE
Add v1 Karpenter disrupted taint to pre-stop hook

### LINKED ISSUE #2158: karpenter disruption taint isn't correct for karpenter v1
/kind bug

**What happened?**

EBS CSI driver does not clean up after itself when node is terminating due to karpenter disruption after v1.0.1 karpenter upgrade

**What you expected to happen?**

EBS CSI driver removes all volume attachments when terminating due to node termination

**How to reproduce it (as minimally and precisely as possible)?**

Have an EBS volume attached, and don't configure any tolerations on the EBS CSI driver (so it gets terminated before the node terminates)

**Anything else we need to know?**:

The taint during karpenter disruption is:
```
  taints:
  - effect: NoSchedule
    key: karpenter.sh/disrupted
 ```
 
 But the `IsDisruptedTaint` looks for `disrupting` rather than `disrupted` because it's using the v1beta1 API rather than the v1 API.

https://github.com/kubernetes-sigs/karpenter/blob/b69e975128ac9a511542f9f1d245a6d4c3f91112/pkg/apis/v1beta1/taints.go#L28

vs

https://github.com/kubernetes-sigs/karpenter/blob/b69e975128ac9a511542f9f1d245a6d4c3f91112/pkg/apis/v1/taints.go#L28

**Environment**
- Kubernetes version (use `kubectl version`):
```
Client Version: v1.31.1
Kustomize Version: v5.4.2
Server Version: v1.30.3-eks-a18cd3a
```
- Driver version: v1.35.0

### PR DESCRIPTION
**Pull request by AndrewSirenko**
**Is this a bug fix or adding new feature?**
Bug fix

**What is this PR about? / Why do we need it?**
Fixes #2158 

Our pre-stop hook attempts to keep the EBS CSI Driver node plugin alive while node is being drained so that it can Unstage/Unpublish all volumes and report that to Kubelet before node termination. Karpenter v1 changed their disruption taint, so we need to update our pre-stop hook to check for it. 

We also remove the dependency on Karpenter because we are compatible with multiple versions of it, and switch to just checking taint key instead of taint key and effect in case these auto-scalars switch to a `NoExecute` effect.

**What testing is done?** 
New unit tests. 

TODO test with karpenter v1.

### CODE REVIEW

**Comment by AndrewSirenko** [comment]
/retest

windows flake for volume expansion

**Review by torredil — APPROVED** [review]

**Review by ElijahQuinones — APPROVED** [review]

**Review by ConnorJC3 — CHANGES_REQUESTED** [review]
One minor change otherwise lgtm

File: cmd/hooks/prestop.go [hunk 1]
```diff
@@ -44,6 +43,16 @@ If the PreStop hook hangs during its execution, the driver node pod will be forc
 */
 
 const clusterAutoscalerTaint = "ToBeDeletedByClusterAutoscaler"
+const v1KarpenterTaint = "karpenter.sh/disrupted"
+const v1beta1KarpenterTaint = "karpenter.sh/disruption"
+
+// drainTaints includes taints used by K8s or autoscalers that signify node draining or pod eviction
+var drainTaints = map[string]bool{
```
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply cmd/hooks/prestop.go:50]: I actually saw this go hack online a couple times but I personally thought it is less readable which is why I decided against it. 

I don't think saving a few bytes of RAM on pre-stop hook is not worth the readability hit for such a small map. (E.g. map of strings-> booleans is self documenting versus map of string -> empty structs might not be)

But if the team would rather have it changed I'm open to it. @ElijahQuinones @torredil.

[the remark this answers is not in the document]
Comment (ElijahQuinones) [reply cmd/hooks/prestop.go:50]: I don't have a hard stance ether way but maybe we can go in the middle Connors change with a comment to make it more readable ?

[the remark this answers is not in the document]
Comment (ConnorJC3) [reply cmd/hooks/prestop.go:50]: It's _more_ readable - the existing code makes you wonder what the `true` means, and you have to read elsewhere to find out it's entirely arbitrary (`false` would do the same thing).

Using `struct{}` leaves no room for thinking switching to `false` would somehow change the behavior or remove the value from the considered taints. It also makes this easy to perform error a syntax error: `if isDrainTaint = drainTaints[someValue]; isDrainTaint {`, which is valid under the bool setup.

Finally, and probably most importantly: `struct{}` is how we already do this pattern everywhere else in our code:
- https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/eaabf29cf318e0135ed78659c3cf68e3efd71af5/pkg/driver/constants.go#L191-L227
- https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/eaabf29cf318e0135ed78659c3cf68e3efd71af5/pkg/cloud/interface.go#L40
- https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/eaabf29cf318e0135ed78659c3cf68e3efd71af5/pkg/cloud/cloud.go#L184-L193
- https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/eaabf29cf318e0135ed78659c3cf68e3efd71af5/pkg/cloud/volume_limits.go#L55-L71

[the remark this answers is not in the document]
Comment (ConnorJC3) [reply cmd/hooks/prestop.go:50]: I have no idea what that reddit post is, have never read it, and don't think we should be making decisions based on random reddit comments one way or the other.

>  For bool values you can do a simple m[k] which is less to type and is easier to understand.

The problem with this is that it makes `ok := map[key]` and `_, ok := map[key]` both valid checks that are arguably correct but mean different things - the bool stores redundant information and it isn't clear to the reader what it represents or what a `false` value would mean - making the code less readable.

[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply cmd/hooks/prestop.go]: Thank you for your feedback, I have changed the PR accordingly

**Review by ConnorJC3 — APPROVED** [review]
/lgtm

**Comment by AndrewSirenko** [comment]
@ElijahQuinones  Not a flake, need to reformat the module import

**Pull request merged**